File: games/tbamud/shops.py
"""Shop system — buy, sell, list, appraise."""

from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from core.engine import Engine
    from core.session import Session


# Item type names for display
ITEM_TYPE_NAMES = {
    1: "빛", 2: "두루마리", 3: "지팡이", 4: "막대기", 5: "무기",
    6: "방화", 7: "투사체", 8: "보물", 9: "갑옷", 10: "물약",
    11: "쓸모없음", 12: "기타", 13: "쓰레기", 14: "컨테이너",
    15: "메모", 16: "음료", 17: "열쇠", 18: "음식",
    19: "돈", 20: "배", 21: "샘물",
}


def register(engine: Engine) -> None:
    engine.register_command("buy", do_buy, korean="사")
    engine.register_command("sell", do_sell, korean="팔")
    engine.register_command("list", do_list, korean="목록")
    engine.register_command("appraise", do_appraise, korean="감정")


def _find_shop(engine: Engine, session) -> tuple:
    """Find shop in current room. Returns (shop, keeper_mob) or (None, None)."""
    char = session.character
    if not char:
        return None, None

    room = engine.world.get_room(char.room_vnum)
    if not room:
        return None, None

    # Find shopkeeper NPC in room
    for mob in room.characters:
        if mob.is_npc and mob.proto.vnum in engine.world.shops:
            return engine.world.shops[mob.proto.vnum], mob

    return None, None


def _is_open(shop, hour: int = 12) -> bool:
    """Check if shop is open at given hour."""
    if shop.open1 <= hour < shop.close1:
        return True
    if shop.open2 <= hour < shop.close2:
        return True
    return False


async def do_buy(session, args: str) -> None:
    """Buy an item from a shop."""
    engine = session.engine
    shop, keeper = _find_shop(engine, session)
    if not shop:
        await session.send_line("여기에는 상점이 없습니다.")
        return

    if not _is_open(shop):
        await session.send_line(f"{keeper.name}이(가) '지금은 영업시간이 아닙니다.'라고 말합니다.")
        return

    if not args:
        await session.send_line(f"{keeper.name}이(가) '뭘 사시겠습니까?'라고 말합니다.")
        return

    char = session.character
    target_kw = args.strip().lower()

    # Search shop's selling items
    for item_vnum in shop.selling_items:
        proto = engine.world.item_protos.get(item_vnum)
        if not proto:
            continue
        if target_kw in proto.keywords.lower():
            price = int(proto.cost * shop.profit_buy)
            if char.gold < price:
                await session.send_line(
                    f"{keeper.name}이(가) '그건 {price} 골드입니다. 돈이 부족합니다.'라고 말합니다."
                )
                return
            # Create item instance
            obj = engine.world.create_obj(proto.vnum)
            char.gold -= price
            char.inventory.append(obj)
            await session.send_line(
                f"{{bright_yellow}}{proto.short_description}을(를) {price} 골드에 구입했습니다.{{reset}}"
            )
            return

    # Also check keeper's inventory
    for obj in keeper.inventory:
        if target_kw in obj.proto.keywords.lower():
            price = int(obj.proto.cost * shop.profit_buy)
            if char.gold < price:
                await session.send_line(
                    f"{keeper.name}이(가) '돈이 부족합니다.'라고 말합니다."
                )
                return
            keeper.inventory.remove(obj)
            char.gold -= price
            char.inventory.append(obj)
            await session.send_line(
                f"{{bright_yellow}}{obj.name}을(를) {price} 골드에 구입했습니다.{{reset}}"
            )
            return

    await session.send_line(f"{keeper.name}이(가) '그런 물건은 없습니다.'라고 말합니다.")


async def do_sell(session, args: str) -> None:
    """Sell an item to a shop."""
    engine = session.engine
    shop, keeper = _find_shop(engine, session)
    if not shop:
        await session.send_line("여기에는 상점이 없습니다.")
        return

    if not _is_open(shop):
        await session.send_line(f"{keeper.name}이(가) '영업시간이 아닙니다.'라고 말합니다.")
        return

    if not args:
        await session.send_line(f"{keeper.name}이(가) '뭘 파시겠습니까?'라고 말합니다.")
        return

    char = session.character
    target_kw = args.strip().lower()

    for obj in char.inventory:
        if target_kw in obj.proto.keywords.lower():
            price = int(obj.proto.cost * shop.profit_sell)
            price = max(1, price)
            char.inventory.remove(obj)
            char.gold += price
            keeper.inventory.append(obj)
            await session.send_line(
                f"{{bright_yellow}}{obj.name}을(를) {price} 골드에 판매했습니다.{{reset}}"
            )
            return

    await session.send_line(f"{keeper.name}이(가) '그런 물건을 가지고 있지 않습니다.'라고 말합니다.")


async def do_list(session, args: str) -> None:
    """List items for sale in a shop."""
    engine = session.engine
    shop, keeper = _find_shop(engine, session)
    if not shop:
        await session.send_line("여기에는 상점이 없습니다.")
        return

    if not _is_open(shop):
        await session.send_line(f"{keeper.name}이(가) '영업시간이 아닙니다.'라고 말합니다.")
        return

    lines = [f"{{bright_cyan}}{keeper.name}의 상점 물품 목록:{{reset}}"]

    # Permanent stock items
    idx = 1
    for item_vnum in shop.selling_items:
        proto = engine.world.item_protos.get(item_vnum)
        if not proto:
            continue
        price = int(proto.cost * shop.profit_buy)
        type_name = ITEM_TYPE_NAMES.get(proto.item_type, "기타")
        lines.append(f"  {idx}. {proto.short_description} [{type_name}] — {price} 골드 (무한)")
        idx += 1

    # Keeper's inventory
    for obj in keeper.inventory:
        price = int(obj.proto.cost * shop.profit_buy)
        type_name = ITEM_TYPE_NAMES.get(obj.proto.item_type, "기타")
        lines.append(f"  {idx}. {obj.name} [{type_name}] — {price} 골드")
        idx += 1

    if idx == 1:
        lines.append("  (판매 중인 물건이 없습니다)")

    await session.send_line("\r\n".join(lines))


async def do_appraise(session, args: str) -> None:
    """Appraise an item's sell value."""
    engine = session.engine
    shop, keeper = _find_shop(engine, session)
    if not shop:
        await session.send_line("여기에는 상점이 없습니다.")
        return

    if not args:
        await session.send_line("뭘 감정하시겠습니까?")
        return

    char = session.character
    target_kw = args.strip().lower()

    for obj in char.inventory:
        if target_kw in obj.proto.keywords.lower():
            price = int(obj.proto.cost * shop.profit_sell)
            price = max(1, price)
            await session.send_line(
                f"{keeper.name}이(가) '{obj.name}은(는) {price} 골드에 사겠습니다.'라고 말합니다."
            )
            return

    await session.send_line("그런 물건을 가지고 있지 않습니다.")

File: games/tbamud/lua/commands/shops.lua
-- shops.lua — tbaMUD shop commands (buy, sell, list, appraise)

local ITEM_TYPE_NAMES = {
    [1]="빛", [2]="두루마리", [3]="지팡이", [4]="막대기", [5]="무기",
    [6]="방화", [7]="투사체", [8]="보물", [9]="갑옷", [10]="물약",
    [11]="쓸모없음", [12]="기타", [13]="쓰레기", [14]="컨테이너",
    [15]="메모", [16]="음료", [17]="열쇠", [18]="음식",
    [19]="돈", [20]="배", [21]="샘물",
}

-- ── buy ─────────────────────────────────────────────────────────

register_command("buy", function(ctx, args)
    local shop, keeper = ctx:find_shop()
    if not shop then
        ctx:send("여기에는 상점이 없습니다.")
        return
    end
    if not args or args == "" then
        ctx:send(keeper.name .. "이(가) '뭘 사시겠습니까?'라고 말합니다.")
        return
    end

    local ch = ctx.char
    if not ch then return end
    local target_kw = args:lower()

    -- Search shop's permanent selling items
    local selling = shop.selling_items
    if selling then
        for i = 1, #selling do
            local item_vnum = selling[i]
            if item_vnum then
                local obj = ctx:create_obj(item_vnum)
                if obj and obj.proto.keywords:lower():find(target_kw, 1, true) then
                    local price = math.floor(obj.proto.cost * shop.profit_buy)
                    if ch.gold < price then
                        ctx:send(keeper.name .. "이(가) '그건 " .. price ..
                                 " 골드입니다. 돈이 부족합니다.'라고 말합니다.")
                        return
                    end
                    ch.gold = ch.gold - price
                    ctx:obj_to_char(obj, ch)
                    ctx:send("{bright_yellow}" .. obj.proto.short_description ..
                             "을(를) " .. price .. " 골드에 구입했습니다.{reset}")
                    return
                end
            end
        end
    end

    -- Check keeper's inventory
    local keeper_inv = ctx:get_char_inventory(keeper)
    for i = 1, #keeper_inv do
        local obj = keeper_inv[i]
        if obj and obj.proto.keywords:lower():find(target_kw, 1, true) then
            local price = math.floor(obj.proto.cost * shop.profit_buy)
            if ch.gold < price then
                ctx:send(keeper.name .. "이(가) '돈이 부족합니다.'라고 말합니다.")
                return
            end
            ctx:obj_from_char(obj)
            ch.gold = ch.gold - price
            ctx:obj_to_char(obj, ch)
            ctx:send("{bright_yellow}" .. obj.name .. "을(를) " .. price ..
                     " 골드에 구입했습니다.{reset}")
            return
        end
    end

    ctx:send(keeper.name .. "이(가) '그런 물건은 없습니다.'라고 말합니다.")
end, "사")

-- ── sell ────────────────────────────────────────────────────────

register_command("sell", function(ctx, args)
    local shop, keeper = ctx:find_shop()
    if not shop then
        ctx:send("여기에는 상점이 없습니다.")
        return
    end
    if not args or args == "" then
        ctx:send(keeper.name .. "이(가) '뭘 파시겠습니까?'라고 말합니다.")
        return
    end

    local ch = ctx.char
    if not ch then return end
    local obj = ctx:find_obj_inv(args)
    if not obj then
        ctx:send(keeper.name .. "이(가) '그런 물건을 가지고 있지 않습니다.'라고 말합니다.")
        return
    end

    local price = math.floor(obj.proto.cost * shop.profit_sell)
    if price < 1 then price = 1 end
    ctx:obj_from_char(obj)
    ch.gold = ch.gold + price
    ctx:send("{bright_yellow}" .. obj.name .. "을(를) " .. price .. " 골드에 판매했습니다.{reset}")
end, "팔")

-- ── list ────────────────────────────────────────────────────────

register_command("list", function(ctx, args)
    local shop, keeper = ctx:find_shop()
    if not shop then
        ctx:send("여기에는 상점이 없습니다.")
        return
    end

    local lines = {"{bright_cyan}" .. keeper.name .. "의 상점 물품 목록:{reset}"}
    local idx = 1

    -- Permanent stock items
    local selling = shop.selling_items
    if selling then
        for i = 1, #selling do
            local item_vnum = selling[i]
            if item_vnum then
                local obj = ctx:create_obj(item_vnum)
                if obj then
                    local price = math.floor(obj.proto.cost * shop.profit_buy)
                    local type_name = ITEM_TYPE_NAMES[obj.proto.item_type] or "기타"
                    table.insert(lines, "  " .. idx .. ". " ..
                                 obj.proto.short_description .. " [" .. type_name ..
                                 "] — " .. price .. " 골드 (무한)")
                    idx = idx + 1
                end
            end
        end
    end

    -- Keeper's inventory
    local keeper_inv = ctx:get_char_inventory(keeper)
    for i = 1, #keeper_inv do
        local obj = keeper_inv[i]
        if obj then
            local price = math.floor(obj.proto.cost * shop.profit_buy)
            local type_name = ITEM_TYPE_NAMES[obj.proto.item_type] or "기타"
            table.insert(lines, "  " .. idx .. ". " .. obj.name ..
                         " [" .. type_name .. "] — " .. price .. " 골드")
            idx = idx + 1
        end
    end

    if idx == 1 then
        table.insert(lines, "  (판매 중인 물건이 없습니다)")
    end
    ctx:send(table.concat(lines, "\r\n"))
end, "목록")

-- ── appraise ────────────────────────────────────────────────────

register_command("appraise", function(ctx, args)
    local shop, keeper = ctx:find_shop()
    if not shop then
        ctx:send("여기에는 상점이 없습니다.")
        return
    end
    if not args or args == "" then
        ctx:send("뭘 감정하시겠습니까?")
        return
    end

    local obj = ctx:find_obj_inv(args)
    if not obj then
        ctx:send("그런 물건을 가지고 있지 않습니다.")
        return
    end

    local price = math.floor(obj.proto.cost * shop.profit_sell)
    if price < 1 then price = 1 end
    ctx:send(keeper.name .. "이(가) '" .. obj.name .. "은(는) " .. price ..
             " 골드에 사겠습니다.'라고 말합니다.")
end, "감정")
